File: apps/android/app/src/main/java/io/github/jgalego/ephemeral/Transport.kt
package io.github.jgalego.ephemeral

import org.json.JSONArray
import java.io.BufferedReader
import java.net.HttpURLConnection
import java.net.URL
import javax.net.ssl.HttpsURLConnection

/**
 * One HTTPS round trip, performed by this app rather than by the engine.
 *
 * Ephemeral opens no sockets on a phone and brings no HTTP stack, so TLS and
 * everything around it stay with the platform. This interface is the seam; the
 * engine calls [send] whenever it needs the model.
 */
interface Transport {
    /**
     * Sends [body] to [endpoint] by [method] with exactly the headers in
     * [headersJson], returning the response body or null on any failure.
     *
     * [method] is `GET` or `POST` and is to be used as given. [body] is empty
     * for a `GET` and none is to be sent with one. The version of this that
     * always POSTed sent the models listing to `/v1/models` as a POST, which
     * OpenAI refuses with an empty body — so the connection test, the one call
     * a person makes before spending anything, failed with a JSON parse error
     * while generation against the same service worked seconds later.
     *
     * [headersJson] is `[{"name":…,"value":…}, …]` — the complete set the
     * provider composed, credential included. This app sets those and adds
     * nothing: which headers a service wants is the provider's knowledge, and
     * the version of this that wrote Anthropic's headers here is the reason a
     * phone could not be pointed at anything else.
     *
     * [status] is a one-element array to write the HTTP status into, since Java
     * has no out-parameters. Leave it zero if there is none — a generated
     * application allowed to reach a service sees this number, and an invented
     * 200 is one it might branch on.
     *
     * Called from a background thread the engine owns. It must not throw: an
     * exception here crosses back into native code, where there is nothing that
     * can act on it.
     */
    fun send(
        method: String,
        endpoint: String,
        headersJson: String,
        body: String,
        status: IntArray,
    ): String?
}

/** The platform's HTTPS stack, and nothing on top of it. */
internal object Https : Transport {
    private const val CONNECT_TIMEOUT = 30_000
    private const val READ_TIMEOUT = 180_000

    override fun send(
        method: String,
        endpoint: String,
        headersJson: String,
        body: String,
        status: IntArray,
    ): String? {
        var connection: HttpURLConnection? = null
        return try {
            // Plain HTTP would put the credential on the wire in clear text, so
            // it is refused here rather than negotiated.
            val opened = URL(endpoint).openConnection()
            if (opened !is HttpsURLConnection) return null
            connection = opened

            opened.requestMethod = method
            opened.connectTimeout = CONNECT_TIMEOUT
            opened.readTimeout = READ_TIMEOUT
            // Setting this at all makes the request a POST, whatever
            // requestMethod was told — HttpURLConnection decides the method
            // from whether there is a body to write. So it is set only when
            // there is one.
            val sends = method.equals("POST", ignoreCase = true)
            opened.doOutput = sends
            // Whatever the provider asked for, and only that.
            val headers = JSONArray(headersJson)
            for (index in 0 until headers.length()) {
                val header = headers.optJSONObject(index) ?: continue
                opened.setRequestProperty(header.optString("name"), header.optString("value"))
            }

            if (sends) {
                opened.outputStream.use { it.write(body.toByteArray(Charsets.UTF_8)) }
            }

            // Read once and reported, because asking for it again is another
            // trip through the connection's own state machine.
            val code = opened.responseCode
            status[0] = code

            // The error body is returned too, not swallowed: when a provider
            // refuses, its own words are the most useful thing a person can be
            // shown, and the engine already knows how to read them.
            val stream = if (code in 200..299) opened.inputStream else opened.errorStream
            stream?.bufferedReader(Charsets.UTF_8)?.use(BufferedReader::readText)
        } catch (failure: Exception) {
            null
        } finally {
            connection?.disconnect()
        }
    }
}

File: apps/ios/Sources/Ephemeral/Transport.swift
// The HTTPS this application performs on the engine's behalf.
//
// Ephemeral does not open sockets and brings no HTTP stack (ADR-0017). That is
// not a limitation being worked around: it is what makes generating on iOS
// possible at all. The desktop transport hands a request to `curl`, and iOS
// does not let an application spawn a process — so the host supplies the
// transport, and TLS, certificate policy and background behaviour stay with the
// platform that already has opinions about them.
//
// Everything the model is sent leaves this file. Nothing else in the
// application makes a network request.

import Ephemeral
import Foundation

enum Transport {

    /// Sends one request, and blocks until it answers.
    ///
    /// A C function pointer cannot capture anything, so this is a plain
    /// closure with no context — everything it needs arrives as an argument.
    /// Blocking is correct here: the engine's call is synchronous, and it is
    /// already running off the main thread on the engine's own queue.
    static let send: EphemeralHttpSend = { _, method, endpoint, headersJson, body in
        guard let method, let endpoint, let headersJson, let body,
              let url = URL(string: String(cString: endpoint))
        else { return nil }

        // The method the provider asked for, not an assumption. Listing what a
        // service has is a GET, and the version of this that always POSTed sent
        // that to `/v1/models` — which OpenAI refuses with an empty body, so
        // the connection test failed while generation worked.
        let verb = String(cString: method)
        var request = URLRequest(url: url)
        request.httpMethod = verb
        if verb == "POST" {
            request.httpBody = Data(String(cString: body).utf8)
        }

        // Exactly the headers the provider composed, and nothing added here.
        // Which headers a service wants is the provider's knowledge; this file
        // used to write Anthropic's three, which is why a phone could not be
        // pointed at any other service however it was configured.
        let headers = try? JSONSerialization.jsonObject(
            with: Data(String(cString: headersJson).utf8)
        )
        for header in headers as? [[String: String]] ?? [] {
            guard let name = header["name"], let value = header["value"] else { continue }
            request.setValue(value, forHTTPHeaderField: name)
        }

        // Generation is a model request and takes as long as one takes. The
        // default sixty seconds is not enough, and a timeout in the middle of
        // writing an application reads to a person as the application failing.
        request.timeoutInterval = 300

        let waiting = DispatchSemaphore(value: 0)
        var reply: String?

        URLSession.shared.dataTask(with: request) { data, _, _ in
            reply = data.flatMap { String(data: $0, encoding: .utf8) }
            waiting.signal()
        }.resume()

        waiting.wait()

        // `strdup`, because the allocation is ours throughout: Ephemeral copies
        // what it needs immediately and then hands this back to `release`.
        return reply.map { strdup($0) } ?? nil
    }

    /// Frees what `send` returned. The other half of owning the allocation.
    static let release: EphemeralHttpFree = { _, response in
        free(response)
    }
}
